### ISSUE TITLE
Let the user enable language features on arbitrary yaml config

### ISSUE THREAD

**Issue by inliquid** [vscode-kubernetes-tools/vscode-kubernetes-tools#1044]
Say I have a number of `*.yaml` files somewhere in a repo which is later getting assembled in k8s configs by CI/CD pipeline. Which is not uncommon. I would like to have a documentation on hover, validation, and other features, like I do have when work on something like `pod.yaml` or `service.yaml` placed in a root of repository. Maybe that's because of location, maybe because of the fact that these files have all necessary options, no idea which criteria does extension use when decide to serve a particular file. So I want to be able to switch it manually.

The most easy way would be to let choose a language with "Change Language Mode" by VS Code like with other extensions.

**Comment by itowlson** [comment]
The extension should kick in on all `.yaml` or `.yml` files that have a 'Kubernetes-like' header (`apiVersion`, `kind` and `metadata` elements I think).

We left the file type as `YAML` rather than creating a distinct `Kubernetes Manifest` file type so that other YAML extensions would continue to work nicely with k8s files.  (We _did_ introduce a different file type for Helm templates because they're often not valid YAML.)  If your files aren't being picked up then probably we need to look at the detection logic rather than introduce a new file type.

Could you post a sample of a file that isn't working right?  (redacting anything sensitive of course!)

**Comment by inliquid** [comment]
Unfortunately I can't share full config, but I can provide some example of it:
![изображение](https://user-images.githubusercontent.com/6797881/155389859-2144724f-9c94-42b8-8b84-d83bb4539b07.png)

**Comment by itowlson** [comment]
Sorry for not following up on this earlier.  Is it possible for you to share the file extension (and ideally path) and first few lines of your sample please?  It's the file path/extension and the `apiVersion / kind / metadata` triplet that we trigger on.  You can redact the `metadata` values but it would be really useful to see the `apiVersion` and `kind`.  Thanks!

**Comment by inliquid** [comment]
Yep, sure. As I said this file is to be assembled into a working config by CI/CD process, so `apiVersion` and `kind` not present:
![изображение](https://user-images.githubusercontent.com/6797881/159186164-2befcafb-369d-4628-94fe-21f9d2463a20.png)

Path is `.o3/k8s/values_production.yaml`

**Comment by itowlson** [comment]
Ah, okay, it's the lack of `apiVersion` and `kind` that's throwing us off.  We depend on those to know what kind of resource it is (and that it's a k8s resource at all).

It seems like to meet your needs we'd need to do the following:

1. Provide a way to say "this is a k8s resource despite the lack of the standard header" - this would be fairly easy to do through a specially formatted YAML comment, but might be harder to do automagically.  A "Kubernetes Manifest" language could do it but, but, as mentioned, I'm not sure this would work given that we depend on the YAML language type to trigger the Red Hat YAML extension to get involved
2. Provide a way to say "this is the type of k8s resource it is" - if we are using a special YAML comment then we could require the user to specify it there, otherwise not quite sure how to infer this.  A language mode wouldn't solve this.
3. In cases where this special override is provided/detected, _also relax the schema_ so that you don't get errors for the missing mandatory fields that will be added by your CI/CD system - this is probably do-able depending on how we trigger the override, but a bit trickier

This feels like a significant chunk of effort and complexity for a use case that seems quite specific to your environment.  If other users are using this same workflow then it would be worth doing but we don't want to be in the position where we are on the hook for multiple customers' in-house workflows.  And unfortunately this isn't something we have an API for that would allow you to customise the behaviour.

I see a couple of ways forward:

1. We can look at an API that would allow you to write your own VS Code extension that hooked up your documents to the k8s extension and gave them appropriate schemas (derived from the ones the k8s extension would use).
2. You could use "dummy" `apiVersion/kind/metadata` fields, and modify your CI/CD system to strip these off before assembling them into k8s configs.  This would make your documents Just Work not only with the VS Code extension but with any other k8s tooling that cared about resource types.

I'm very happy to discuss further and would love to hear if there are other ways forward or if I'm missing something!
